core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -0,0 +1,2103 @@
+        {
+            StageId stageId = event.getStageId();
+            StageExecution stageExecution = getStageExecution(stageId);
+            event.getPartitionIds().forEach(partitionId -> {
+                Optional<ScheduledTask> scheduledTask = stageExecution.sealPartition(partitionId);
+                scheduledTask.ifPresent(task -> {
+                    if (nodeAcquisitions.containsKey(task)) {
+                        // task is already waiting for node
+                        return;
+                    }
+                    schedulingQueue.addOrUpdate(task, task.getPriority());
+                });
+            });
+        }
+
+        @Override
+        public void onNoMorePartitionsTaskSourceEvent(NoMorePartitionsTaskSourceEvent event)
+        {
+            StageExecution stageExecution = getStageExecution(event.getStageId());
+            stageExecution.noMorePartitions();
+        }
+
+        @Override
+        public void onFailedTaskSourceEvent(FailedTaskSourceEvent event)
+        {
+            StageExecution stageExecution = getStageExecution(event.getStageId());
+            stageExecution.fail(event.getFailure());
+        }
+
+        private StageExecution getStageExecution(StageId stageId)
+        {
+            StageExecution execution = stageExecutions.get(stageId);
+            checkState(execution != null, "stage execution does not exist for stage: %s", stageId);
+            return execution;
+        }
+
+        private static List<SubPlan> sortPlanInTopologicalOrder(SubPlan subPlan)
+        {
+            ImmutableList.Builder<SubPlan> result = ImmutableList.builder();
+            Traverser.forTree(SubPlan::getChildren).depthFirstPreOrder(subPlan).forEach(result::add);
+            return result.build();
+        }
+
+        private boolean shouldDelayScheduling(@Nullable ErrorCode errorCode)
+        {
+            return errorCode == null || errorCode.getType() == INTERNAL_ERROR || errorCode.getType() == EXTERNAL;
+        }
+
+        private ExecutionFailureInfo rewriteTransportFailure(ExecutionFailureInfo executionFailureInfo)
+        {
+            if (executionFailureInfo.getRemoteHost() == null || failureDetector.getState(executionFailureInfo.getRemoteHost()) != GONE) {
+                return executionFailureInfo;
+            }
+
+            return new ExecutionFailureInfo(
+                    executionFailureInfo.getType(),
+                    executionFailureInfo.getMessage(),
+                    executionFailureInfo.getCause(),
+                    executionFailureInfo.getSuppressed(),
+                    executionFailureInfo.getStack(),
+                    executionFailureInfo.getErrorLocation(),
+                    REMOTE_HOST_GONE.toErrorCode(),
+                    executionFailureInfo.getRemoteHost());
+        }
+    }
+
+    private static class StageExecution
+    {
+        private static final int SPECULATIVE_EXECUTION_PRIORITY = 1_000_000_000;
```

Comment:
It is not obvious to me that current code cannot deadlock due to dowstream task consuming resources needed for the upstream tasks. It feels like it comes form the fact that it will never be the case that we start polling from `schedulingQueue` while we put the low prio tasks there but not yet high prio tasks.
But it is not super obvious why is that the case.

Can you elaborate for future code readers about that matter?

Reply:
I'm not sure how much of a sense does it make to schedule downstream tasks (that are speculative) if any of the upstream tasks can be scheduled. Speculative execution is generally needed to improve resource utilization, and if resources can be utilized by non speculative tasks not sure if scheduling speculative tasks is necessary. We can think about it more when we start working on the speculative execution.

Reply:
I mean the case when upstream **cannot** be scheduled because it is more reasource intensive than dowstream task (think binpacking small resource holes with downstream tasks while big upstreasm tasks are still waiting in the queue)

Reply:
Hmm, yeah, you are right. I think we should be carefull with it then and don't allow downstream stages to execute before upstream stages are scheduled.